### DISCUSSION TITLE
IPv4-IPv6 primary differences

### DISCUSSION THREAD

**Discussion by EduardVasilenko** [becarpenter/book6#28] [Ideas]
Hi Brian,
IPv6 basic is started from the IPv6 overview (you have included host-router-node definition).
IMHO: it makes sense to discuss IPv4-IPv6 differences more deeply in this section. Do you agree?
If yes, I could elaborate on the below.

IPv4-IPv6 primary differences:
1.	32bits address size against 64-128bits (depends on SLAAC versus DHCP for address acquisition, other bits are used for privacy or other purposes)
2.	NAT44 is a common solution against NAT66 is discouraged and not specified as a standard, end-to-end-connectivity is the IPv6 value
3.	Only one IPv4 address per interface against many IPv6 addresses as the norm (LLA, ULA, many instances of GUA)
4.	Only centralized DHCPv4 address management against distributed address acquisition of SLAAC
5.	IPv4 variable header against IPv6 fixed header with extension headers when needed
6.	Fragmentation in the basic header against fragmentation in the extension header
7.	Link address resolution at L2 against L3 (over LLA IP addresses)
8.	Multicast is not needed for IPv4 itself against Multicast is mandatory for the IPv6 link operation

Eduard

**Comment by becarpenter** [comment]
One goal of course is to normalize the idea of IPv6 as the main protocol and IPv4 as the old legacy. So while I agree that making this sort of comparison list is a good idea, I don't want it to be early in the reading sequence. I think it could fit naturally in Chapter 3: Coexistence with Legacy IPv4. What do you think?

**Comment by EduardVasilenko** [comment]
agree.
But it means that I have to develop this block of text.

**Comment by becarpenter** [comment]
Good! If you add a file to chapter 3 like "IPv4-IPv6 differences.md I can work the magic to add it to the book.

**Comment by EduardVasilenko** [comment]
I could not because:
"Uploads are disabled.
File uploads require push access to this repository."

I could only publish it here.

**Reply by becarpenter** [reply]
I don't really understand that. I thought you could add a file via a PR.

**Reply by EduardVasilenko** [reply]
You are right. My fault. Pull request is posisble.

**Reply by EduardVasilenko** [reply]
You are right. My fault. Pull request is posisble.

**Comment by EduardVasilenko** [comment]
## IPv6 primary differences from IPv4

IPv6 is very similar for transit routers but has some considerable differences on the first hop for hosts as well as for routers.
The primary differences are:
•	The first difference is very desirable and expected, IPv6 has a four times bigger address size (128bits against 32bits). SLAAC is more used on IPv6 than DHCP. SLAAC subnet is 64bits for historical reasons that are fixed in many standards. 2^64 hosts are not possible in one subnet but the address space is reserved even for a smartphone. Hence, it is disputable what is the effective IPv6 address space. It is bigger than 64bits but many IID bits are utilized for privacy and security, not for addressing per ser.
•	NAT44 is a common solution in IPv4 networking.
NAT66 is discouraged by IETF and not specified as a standard. IPv6 end-to-end connectivity is considered a big value.
•	IPv4 has only one address per interface (without special hacks).
Many IPv6 addresses on every interface are the norm. It is not just different address types (LLA, ULA, and GUA) but additionally many instances of GUA and ULA for security or virtualization reasons. The popular ChromOS has seven IPv6 addresses as the minimum. Additionally, the number of IPv6 addresses per interface could almost double in the case of link renumbering.
•	IPv4 has only centralized DHCPv4 address acquisition.
IPv6 has additionally distributed address acquisition by SLAAC which is more adopted. SLAAC considerably changes the logic of the link operation.
•	IPv4 has a complex (many fields) and theoretically variable header that is practically fixed because options are not widely used. 
IPv6 has a simple and fixed header. Additionally, IPv6 could have extension headers that permit unlimited protocol extensibility at the data plane. Many extension headers are already used in limited domains.
•	IPv4 fragmentation is in the basic header and permitted in transit.
IPv6 fragmentation in the extension header and prohibited on the transit.
•	IPv4 address resolution on the link by ARP protocol is at layer 2 (for the ethernet media it is ethernet frame). IPv6 address resolution on the link by ND protocol is at layer 3 (IPv6 packet over LLA or other IP addresses).
•	Multicast is not needed for IPv4 itself.
Multicast is mandatory for the IPv6 link operation. Many ND functions are using multicast. That may create advantages (for Ethernet) and disadvantages (for many types of wireless).
The list above is not comprehensive, but the other differences are probably smaller.


### [<ins>Previous</ins>](Tunnels.md) [<ins>Previous</ins>](Obsolete%20techniques.md) [<ins>Chapter Contents</ins>](3.%20Coexistence%20with%20Legacy%20IPv4.md)

**Comment by becarpenter** [comment]
Forgot to close this when Eduard's text was added to the book.
